docs/add-code-flow.md

```diff
@@ -0,0 +1,97 @@
+https://github.com/ipfs/go-ipfs/blob/master/core/commands/add.go#L208-L213
+
+The goal of this document is to capture the code flow for adding a file (see the `coreapi` package) using the IPFS CLI, in the process exploring some datastructures and packages like `ipld.Node` (aka `dagnode`, `FSNode`, `MFS`, etc.
+
+# `UnixfsAPI.Add()`
+*Entrypoint into the `Unixfs` package*
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreapi/unixfs.go#L78-L86
+
+The `UnixfsAPI.Add()` acts on the input data or files, to build a _merkledag_ node (in essence it is the entire tree represented by the root node) and adds it to the _blockstore_.
+Within the function, a new `Adder` is created with the configured `Blockstore` and __DAG service__`. 
+
+
+# `Adder.AddAllAndPin(files.File)`
+*Entrypoint*
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L427-L431
+
+The more interesting stuff happens in the `Adder.AddAllAndPin(files)` function. 
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L467-L476
+
+Let's focus on the simple case of a single file, handled by `Adder.addFileNode(path, file)` which redirects the case to `Adder.addFile(pathm files.File)`. 
+
+## `Adder.addFile(path, files.File)`
+*Create the _DAG_ and add to `MFS`*
+
+The `addFile(file)` method is responsible taking the data and converting it into a __DAG__ tree, followed by adding the root of the DAG tree in to the `MFS`.
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L508-L521
+
+There are two main methods to focus on -
+
+### `Adder.add(io.Reader)`
+*Create and return the **root** __DAG__ node*
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L115-L137
+
+This method converts the input _data_ (`io.Reader`) to a `DAG` tree. This is done by splitting the data into _chunks_ using the `Chunker` **(HELP: elaborate on chunker types ?)** and organizing them in a `DAG` (with a *trickle* or *balanced* layout). The method returns the **root** of the __DAG__, formatted as an `ipld.Node`.
```

Comment:
Let's just mention the balanced builder, we can point to the code documentation which has some illustrative diagrams:

https://github.com/ipfs/go-unixfs/blob/6b769632e7eb8fe8f302e3f96bf5569232e7a3ee/importer/balanced/builder.go

(`go-unixfs` version associated with `go-ipfs` 4.18).